samza-operator/src/main/java/org/apache/samza/operators/impl/OperatorImpls.java

```diff
@@ -0,0 +1,124 @@
+   */
+  private static final Map<OperatorSpec, OperatorImpl> OPERATOR_IMPLS = new ConcurrentHashMap<>();
+
+  /**
+   * Traverses the DAG of {@link OperatorSpec} starting from the provided {@link MessageStreamImpl}, creates the
+   * corresponding DAG of {@link OperatorImpl}s, and returns its root {@link NoOpOperatorImpl} node.
+   *
+   * @param source  the input {@link MessageStreamImpl} to instantiate {@link OperatorImpl}s for
+   * @param <M>  type of the {@link Message} in the input {@link MessageStreamImpl}
+   * @param context  the {@link TaskContext} required to instantiate operators
+   * @return  root node for the {@link OperatorImpl} DAG.
+   */
+  public static <M extends Message> NoOpOperatorImpl<M> createOperatorImpls(MessageStreamImpl<M> source, TaskContext context) {
+    // since the source message stream might have multiple operator specs registered on it,
+    // create a new root node as a single point of entry for the DAG.
+    NoOpOperatorImpl<M> rootOperator = new NoOpOperatorImpl<>();
+    // create the pipeline/topology starting from the source
+    source.getRegisteredOperatorSpecs().forEach(registeredOperator -> {
+        // pass in the source and context s.t. stateful stream operators can initialize their stores
+        OperatorImpl<M, ? extends Message> operatorImpl =
+            createAndRegisterOperatorImpl(registeredOperator, source, context);
+        rootOperator.registerNextOperator(operatorImpl);
+      });
+    return rootOperator;
+  }
+
+  /**
+   * Helper method to recursively traverse the {@link OperatorSpec} DAG and instantiate and link the corresponding
+   * {@link OperatorImpl}s.
+   *
+   * @param operatorSpec  the operatorSpec registered with the {@code source}
+   * @param source  the source {@link MessageStreamImpl}
+   * @param context  the context of the task
+   * @return  the operator implementation for the operatorSpec
+   */
+  private static <M extends Message> OperatorImpl<M, ? extends Message> createAndRegisterOperatorImpl(OperatorSpec operatorSpec,
+      MessageStream source, TaskContext context) {
+    if (!OPERATOR_IMPLS.containsKey(operatorSpec)) {
+      OperatorImpl<M, ? extends Message> operatorImpl = createOperatorImpl(operatorSpec);
+      if (OPERATOR_IMPLS.putIfAbsent(operatorSpec, operatorImpl) == null) {
+        // this is the first time we've added the operatorImpl corresponding to the operatorSpec,
+        // so traverse and initialize and register the rest of the DAG.
+        MessageStream<? extends Message> outStream = operatorSpec.getOutputStream();
+        Collection<OperatorSpec> registeredSpecs = ((MessageStreamImpl) outStream).getRegisteredOperatorSpecs();
+        registeredSpecs.forEach(registeredSpec -> {
+            OperatorImpl subImpl = createAndRegisterOperatorImpl(registeredSpec, outStream, context);
+            operatorImpl.registerNextOperator(subImpl);
+          });
+        operatorImpl.init(source, context);
```

Comment:
I think that we discussed this one before. We can remove the init() method from OperatorImpl if we pass source and context as parameters to the constructor of OperatorImpl. i.e. change createOperatorImpl(operatorSpec) to createOperatorImpl(operatorSpec, source, context). Or you are planning to fix it in another PR?